src/index.ts

```diff
@@ -1,10 +1,37 @@
-import type { Params } from 'pinejs-client-core';
+import type {
+	AnyObject,
+	AnyResource,
+	ConstructorParams,
+	ODataOptions,
+	Params,
+	PromiseResultTypes,
+	Resource,
+	RetryParameters,
+} from 'pinejs-client-core';
 import { PinejsClientCore } from 'pinejs-client-core';
 import type * as express from 'express';
 import type { CallbackHandler, Response, Test } from 'supertest';
 import supertest from './supertest';
 import { expect } from './chai';
 import type { UserParam, Overwrite } from './common';
+import type {
+	PickDeferred,
+	PickExpanded,
+} from '@balena/abstract-sql-to-typescript';
+
+type StringKeyOf<T> = keyof T & string;
+type SelectPropsOf<T extends Resource['Read'], U extends ODataOptions<T>> =
+	U['$select'] extends ReadonlyArray<StringKeyOf<T>>
+		? U['$select'][number]
+		: U['$select'] extends StringKeyOf<T>
+			? U['$select']
+			: never;
+type ExpandPropsOf<T extends Resource['Read'], U extends ODataOptions<T>> =
+	U['$expand'] extends ReadonlyArray<StringKeyOf<T>>
+		? U['$expand'][number]
+		: U['$expand'] extends { [key in StringKeyOf<T>]?: any }
+			? StringKeyOf<U['$expand']>
+			: never;
```

Comment:
No, this is the case where `$expand` did not match any type from which we can infer the expanded keys so it is a `never` case - it should not happen and we type it like this to let typescript know that (and if it does happen because someone has gone outside of the typings or similar then typescript will error)